src/sampling.rkt

```diff
@@ -113,8 +114,27 @@
     [else
      hyperrects-analysis]))
 
+(define (preprocessing-<=? alt preprocessing-one preprocessing-two)
+  (<= (errors-score (errors (alt-program alt) (*pcontext*) (*output-repr*) #:processing preprocessing-one))
+      (errors-score (errors (alt-program alt) (*pcontext*) (*output-repr*) #:processing preprocessing-two))))
+
+(define (drop-at ls index)
+  (define-values (front back) (split-at ls index))
+  (append front (rest back)))
+
+(define (remove-unecessary-preprocessing alt preprocessing)
+  (define dropped
+    (for/or ([i (in-range (length preprocessing))])
+      (if (preprocessing-<=? alt (drop-at preprocessing i) preprocessing)
+          (drop-at preprocessing i)
+          #f)))
+  (if dropped
+      (remove-unecessary-preprocessing alt dropped)
+      preprocessing))
+
```

Comment:
This sort of seems unnecessarily complicated. And isn't it O(n^2)? Not that that's going to be a big problem, by why not just test each preprocessing step independently?

Reply:
I forgot to ask you about this- do you know if there will ever be any case where removing a preprocessing step causes another to also not effect the result? I wanted it to be correct since I didn't know.

Reply:
I'm not sure this helps though. It still drops things greedily in left-to-right order, it just redoes a whole bunch of work when it does that. I'd just do the obvious O(n) loop.